File: py/port_scanner.py
import csv
from ipaddress import ip_network
from typing import List

import nmap


def scan_ports(target: str) -> List[tuple]:
    nm = nmap.PortScanner()
    nm.scan(
        hosts=target,
        arguments="-sV",  # TCP scan all ports with version detection
        # hosts=target, arguments="-sT -p 1-65535 -T4"
    )  # TCP scan all ports with aggressive timing
    results = []

    for host in nm.all_hosts():
        for proto in nm[host].all_protocols():
            lport = nm[host][proto].keys()
            for port in lport:
                service = nm[host:str][proto:str][port:str]["state"]
                results.append(
                    (
                        host,
                        port,
                        proto,
                        service.get("name", ""),
                        service.get("state", ""),
                        service.get("product", ""),
                        service.get("extrainfo", ""),
                        service.get("reason", ""),
                        service.get("version", ""),
                        service.get("conf", ""),
                    )
                )

    return results


def write_to_file(results: List[tuple], output_file: str):
    fieldnames = [
        "host",
        "port",
        "protocol",
        "service_name",
        "state",
        "product",
        "extrainfo",
        "reason",
        "version",
        "conf",
    ]

    with open(output_file, "w", newline="") as csvfile:
        writer = csv.writer(csvfile, delimiter=",")
        writer.writerow(fieldnames)
        writer.writerows(results)


def main():
    target = input("Enter the target IP address and subnet (e.g., 192.168.1.0/24 ): ")

    try:
        ip_network(target)
    except ValueError:
        print(f"Invalid target: {target}")
        return

    output_file = f"{target.replace('/', '_')}_port_scan.csv"

    try:
        results = scan_ports(target)
        write_to_file(results, output_file)
        print(f"Scan results saved to {output_file}")
    except nmap.PortScannerError as e:
        print(f"Scan failed: {e}")
    except nmap.PortScannerHostDiscoveryError as e:
        print(f"Scan failed: {e}")
    except nmap.PortScannerTimeout as e:
        print(f"Scan timeout: {e}")
    except Exception as e:
        print(f"An error occurred: {e}")


if __name__ == "__main__":
    main()

File: go/scanner/port_scanner.go
package main

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"

	"github.com/Ullaakut/nmap"
)

func scanHost(host string, portRange string) [][]string {
	scanner, err := nmap.NewScanner(
		nmap.WithTargets(host),
		nmap.WithPorts(portRange),
		nmap.WithServiceInfo(),
		nmap.WithMinRate(1000),
	)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating scanner on host %s: %v\n", host, err)
		return nil
	}

	result, _, err := scanner.Run()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error scanning host %s: %v\n", host, err)
		return nil
	}

	var results [][]string

	if len(result.Hosts) > 0 && result.Hosts[0].Status.State == "up" {
		fmt.Printf("Host %s is up\n", host)
		for _, port := range result.Hosts[0].Ports {
			results = append(results, []string{
				host,
				strconv.FormatUint(uint64(port.ID), 10),
				port.Protocol,
				port.Service.Name,
				port.State.String(),
				port.Service.Product,
				port.Service.ExtraInfo,
				port.State.Reason,
				port.Service.Version,
				"",
			})
		}
	}

	return results
}

func scanNetwork(target string, portRange string, maxWorkers int) [][]string {
	ip, ipnet, err := net.ParseCIDR(target)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Invalid target: %s\n", target)
		return nil
	}

	var results [][]string
	var mutex sync.Mutex
	var wg sync.WaitGroup

	jobCh := make(chan string, maxWorkers)

	for i := 0; i < maxWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for host := range jobCh {
				result := scanHost(host, portRange)
				mutex.Lock()
				results = append(results, result...)
				mutex.Unlock()
			}
		}()
	}

	for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); incIP(ip) {
		if !ip.Equal(ipnet.IP) && !ip.Equal(broadcastAddr(ipnet)) {
			jobCh <- ip.String()
		}
	}
	close(jobCh)
	wg.Wait()

	return results
}

func writeToFile(results [][]string, outputFile string) error {
	file, err := os.Create(outputFile)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(bufio.NewWriter(file))
	defer writer.Flush()

	err = writer.Write([]string{
		"host",
		"port",
		"protocol",
		"service_name",
		"state",
		"product",
		"extrainfo",
		"reason",
		"version",
		"conf",
	})
	if err != nil {
		return err
	}

	return writer.WriteAll(results)
}

func main() {
	var target, portRange string
	fmt.Print("Enter the target IP address and subnet (e.g., 192.168.1.0/24): ")
	fmt.Scanln(&target)
	fmt.Print("Enter the port range to scan (e.g., 1-1000): ")
	fmt.Scanln(&portRange)

	outputFile := strings.ReplaceAll(target, "/", "_") + "_port_scan.csv"

	results := scanNetwork(target, portRange, 50)
	if err := writeToFile(results, outputFile); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing to file: %v\n", err)
		return
	}
	fmt.Printf("Scan completed, %d open ports found\n", len(results))
	fmt.Printf("Scan results saved to %s\n", outputFile)
}

func incIP(ip net.IP) {
	for j := len(ip) - 1; j >= 0; j-- {
		ip[j]++
		if ip[j] > 0 {
			break
		}
	}
}

func broadcastAddr(n *net.IPNet) net.IP {
	bc := make(net.IP, len(n.IP))
	copy(bc, n.IP)
	offset := len(bc) - len(n.Mask)
	for i := range bc {
		if i >= offset {
			bc[i] |= ^n.Mask[i-offset]
		}
	}
	return bc
}
